charmcraft/config.py

```diff
@@ -14,171 +14,282 @@
-        prime = parts.get("bundle", {}).get("prime", [])
-        return cls(prime)
+    prime: List[pydantic.StrictStr] = []
 
+    @pydantic.validator("prime", each_item=True)
+    def validate_relative_paths(cls, prime):
+        """Verify relative paths are used in prime."""
+        return check_relative_paths(prime)
 
-@attr.s(kw_only=True, frozen=True)
-class Project:
-    """Configuration for all project-related options, used internally."""
 
-    dirpath = attr.ib(default=None)
-    config_provided = attr.ib(default=None)
-    started_at = attr.ib(default=None)
+class Parts(pydantic.BaseModel, frozen=True, validate_all=True):
+    """Definition of parts to build."""
 
+    __root__: Dict[pydantic.StrictStr, Part] = {}
 
-@attr.s(kw_only=True, frozen=True)
-class Config:
-    """Root of all the configuration."""
+    def get(self, part_name) -> Part:
+        """Get part by name.
 
-    charmhub = attr.ib(default={}, converter=CharmhubConfig.from_dict)
-    parts = attr.ib(default={}, converter=BasicPrime.from_dict)
-    type = attr.ib(default=None)
+        :returns: Part if exists, None if not.
+        """
+        try:
+            return self.__root__[part_name]
+        except KeyError:
+            return None
 
-    # this item is provided by the code itself, not the user, as convenience for the
-    # rest of the code
-    project = attr.ib(default=None)
 
+class CharmhubConfig(
+    pydantic.BaseModel, extra=pydantic.Extra.forbid, frozen=True, validate_all=True
+):
+    """Definition of Charmhub endpoint configuration."""
 
-CONFIG_SCHEMA = {
-    "type": "object",
-    "properties": {
-        "type": {"type": "string", "enum": ["charm", "bundle"]},
-        "charmhub": {
-            "type": "object",
-            "properties": {
-                "api_url": {"type": "string", "format": "url"},
-                "storage_url": {"type": "string", "format": "url"},
-            },
-            "additionalProperties": False,
-        },
-        "parts": {
-            "type": "object",
-            "properties": {
-                "bundle": {
-                    "type": "object",
-                    "properties": {
-                        "prime": {
-                            "type": "array",
-                            "items": {
-                                "type": "string",
-                                "format": "relative_path",
-                            },
-                        },
-                    },
-                },
-            },
-        },
-    },
-    "required": ["type"],
-    "additionalProperties": False,
-}
+    api_url: pydantic.HttpUrl = "https://api.charmhub.io"
+    storage_url: pydantic.HttpUrl = "https://storage.snapcraftcontent.com"
+
+
+class Project(
+    pydantic.BaseModel, extra=pydantic.Extra.forbid, frozen=True, validate_all=True
+):
+    """Internal-only project configuration."""
+
+    dirpath: pydantic.DirectoryPath
+    content: Dict[str, Any] = {}
```

Comment:
Why we need this `content` here?

Reply:
It was already there https://github.com/canonical/charmcraft/pull/323/files#diff-0657ee63e8720cf85a4fc089151f357cd8309bf35f2213c20f2845f9cd67683aL211  - looks unused so I removed it

Reply:
FTR, there is no "content" attribute in the current (as in `master` branch) config.